### PR TITLE
Display when a repo was archived

### PR DESCRIPTION
**Pull request by JakobDev**
This adds the date a repo is archived to Gitea and shows it in the UI and API. A feature, that GitHub has been [introduced recently](https://github.blog/changelog/2022-11-23-repository-archive-date-now-shown-in-ui/).

I currently don't know how to correctly deal with the Date in the template, as different languages have different Ways of writing a date.

grafik

### CODE REVIEW

**Comment by silverwind** [comment]
> I currently don't know how to correctly deal with the Date in the template, as different languages have different Ways of writing a date.

This ISO format in the screenshot is pretty universal I'd say.

File: modules/timeutil/timestamp.go
Comment (silverwind) [inline modules/timeutil/timestamp.go]: Does this convert timezone? I would not, we should just store UTC in the database and convert for presentation only.

File: models/migrations/v1_19/v240.go
Comment (delvh) [inline models/migrations/v1_19/v240.go]:
```suggestion
// Copyright 2023 The Gitea Authors. All rights reserved.
```

**Comment by silverwind** [comment]
Lint is failing:

```
+ make lint-backend
go run github.com/golangci/golangci-lint/cmd/golangci-lint@v1.50.1 run

modules/timeutil/timestamp.go:65:9: indent-error-flow: if block ends with a return statement, so drop this else and outdent its block (revive)
	} else {
		timePointer := ts.AsTimeInLocation(setting.DefaultUILocation)
		return &timePointer
	}
```

File: options/locale/locale_en-US.ini
Comment (delvh) [inline options/locale/locale_en-US.ini]: We use snake_case for translation keys.
```suggestion
archive.repo_hint = This repository has been archived on %s. You cannot push, interact with issues, or open pull requests anymore.
```

File: models/migrations/v1_19/v240.go
Comment (delvh) [inline models/migrations/v1_19/v240.go]: Hmm, this doesn't seem like a good migration/architecture:
1. The new attribute must first be added to the database.
2. I don't think we should set `= 0` manually. Either use `DEFAULT 0` in the definition, or don't set it at all when not needed. I think `NULL` is in this case an acceptable value as repos cannot be archived at all.
3. I don't think we should use `0 = Unarchived` if we store `IsArchived` too. 
4. This line is wrong as it shouldn't only use `IS NULL` as condition, but `AND is_archived = true` 
So, something like
```go
err := x.Where("archived_unix IS NULL AND is_archived = ?", true).SetExpr("archived_unix", timeutil.TimeStampNow()).Update(new(Repo))
```
(No guarantee if this actually does what I hope it does)
This way, already archived repos will simply be "archived when the gitea version is upgraded"

File: models/repo/archiver.go
Comment (delvh) [inline models/repo/archiver.go:154]: Hmm, I'm inclined to discard this `else` block.
```suggestion
	}
```

**Comment by yardenshoham** [comment]
> I currently don't know how to correctly deal with the Date in the template, as different languages have different ways of writing a date.

You can utilize the automatic time rendering function as in #21440

File: models/repo/archiver.go
Comment (delvh) [inline models/repo/archiver.go]:
```suggestion
	_, err = db.GetEngine(db.DefaultContext).ID(repo.ID).Cols("is_archived", "archived_unix").NoAutoTime().Update(repo)
```

File: options/locale/locale_en-US.ini
Comment (yardenshoham) [inline options/locale/locale_en-US.ini]: Please use snake case
```suggestion
archive.title_date = This repository has been archived on %s. You can view files and clone it, but cannot push or open issues/pull-requests.
```

**Review by delvh — CHANGES_REQUESTED** [review]

File: templates/repo/home.tmpl
Comment (yardenshoham) [inline templates/repo/home.tmpl]:
```suggestion
					{{.locale.Tr "repo.archive.title_date" .Repository.ArchivedUnix.FormatDate}}
```

File: models/repo/archiver.go
Comment (JakobDev) [reply models/repo/archiver.go:154]: That would mean, that this is not updated when unarchiving a repo. So you end up with Repos that have a archive date but who are not archived.

Comment (delvh) [reply models/repo/archiver.go:154]: As I see it:
We don't care about the archived date as long as the repo isn't archived.
If an unarchived repo has an archived date, the date should simply be ignored.

Comment (JakobDev) [reply models/repo/archiver.go:154]: The date can be archived through the API, so this will be visible to end users

File: modules/timeutil/timestamp.go
Comment (JakobDev) [reply modules/timeutil/timestamp.go]: It is used in the API, so I think it should use UTC

File: models/repo/archiver.go
Comment (delvh) [reply models/repo/archiver.go:154]: Okay, yes, I can understand that concern.
It certainly looks weird if the API returns both `archived_unix != 0` and `is_archived = false`.
But still, the current approach is not a good idea either.
(↑ That is meant regarding my other comments, not this suggestion specifically)

Comment (JakobDev) [reply models/repo/archiver.go:154]: The API currently return `null`, if there is no archive date

File: models/migrations/v1_19/v240.go
Comment (JakobDev) [reply models/migrations/v1_19/v240.go]: I now use `DEFUALT 0`, so this is no longer needed

Comment (delvh) [reply models/migrations/v1_19/v240.go]: Hmm, but this means already archived repos will be shown as "has no archived date", so the go code in my suggestion above is still needed.
Or should already archived repos have that inconsistency?

File: models/repo/archiver.go
Comment (delvh) [reply models/repo/archiver.go:154]: I know, as I said, the comment above is not meant for this suggestion anymore, this one is now resolved.

File: models/migrations/v1_19/v240.go
Comment (JakobDev) [reply models/migrations/v1_19/v240.go]: Tey should have this inconsistency. The Archive Date should say, when I repo is archived. This means it shows when the work on the project has stopped. Setting the upgrade date  as archive date for Repos, which are archived for Years is confusing.

Comment (delvh) [reply models/migrations/v1_19/v240.go]: I know, but I think it is better than not displaying an archived date at all.
It has a benefit:
If we do that, we can be sure that `archived_unix == 0 => is_archived = false`, which we otherwise can't.

Comment (JakobDev) [reply models/migrations/v1_19/v240.go]: For what do we need  `archived_unix == 0 => is_archived = false` ?

File: modules/timeutil/timestamp.go
Comment (silverwind) [reply modules/timeutil/timestamp.go]: Yeah we should prefer UTC on both db and api. Ideally we should convert timezone in JS but that is still a TODO, so current presentation convertion happens in the backend IIRC.

File: models/migrations/v1_19/v240.go
Comment (delvh) [reply models/migrations/v1_19/v240.go]: At the moment: It isn't needed for anything as this PR only introduces it.
However, as I've been continuously taught:
Old assumptions are easily forgotten in large (open-source) projects over the years.
I'm fairly certain that someone will at some point rely on exactly this assumption without knowing that it doesn't apply.

File: modules/timeutil/timestamp.go
Comment (JakobDev) [reply modules/timeutil/timestamp.go]: This is the source of `TimeStampNow()`:
https://github.com/go-gitea/gitea/blob/40ba750c4bf1f3f5f8dff5af57b2db4b600f237f/modules/timeutil/timestamp.go#L34-L39
It is a internal function of Gitea. If it not uses UTC, it might affect other parts of Gitea too.

File: models/migrations/v1_19/v240.go
Comment (JakobDev) [reply models/migrations/v1_19/v240.go]: Why not adding a comment e.g. is zero if repo is not archived or archive date is unknown

Comment (delvh) [reply models/migrations/v1_19/v240.go]: Yup, that works too.

File: modules/structs/repo.go
Comment (lafriks) [inline modules/structs/repo.go]: This requires migration. Probably would good to set archived at value to updated at value for all archived repositories

**Review by lafriks — CHANGES_REQUESTED** [review]

File: modules/structs/repo.go
Comment (JakobDev) [reply modules/structs/repo.go]: The Migration is already done by setting the default to 0.

Comment (delvh) [reply modules/structs/repo.go]: No, this is also what I complained about in https://github.com/go-gitea/gitea/pull/22664#discussion_r1090491414 point 1:
With the current implementation, the column won't be created in the database at all, it will simply be read from at runtime.

Comment (JakobDev) [reply modules/structs/repo.go]: I tested it with SQLite. I removed the column, started Gitea and closed it. The Column was there with 0 as value.

File: models/repo/repo.go
Comment (delvh) [inline models/repo/repo.go]:
```suggestion
	ArchivedUnix timeutil.TimeStamp `xorm:"DEFAULT 0"`
```

File: modules/structs/repo.go
Comment (lafriks) [reply modules/structs/repo.go]: column will be added also without migration but it can create problems in future migrations if they will use this column.

https://github.com/go-gitea/gitea/blob/main/CONTRIBUTING.md#design-guideline

File: web_src/js/features/repo-home.js
Comment (yardenshoham) [inline web_src/js/features/repo-home.js]: can you reuse
https://github.com/go-gitea/gitea/blob/b5b3e0714e624cea3ce4d5368aa1266f7639d0eb/web_src/js/features/formatting.js#L5
?

Maybe `export` here?

File: modules/structs/repo.go
Comment (JakobDev) [reply modules/structs/repo.go]: So I should run
```sql
ALTER TABLE repository
ADD COLUMN archived_unix INTEGER DEFAULT 0;
```
during migration?

File: templates/repo/home.tmpl
Comment (delvh) [inline templates/repo/home.tmpl:57]: Should we perhaps rename that to `repo.archive_no_date`?

File: modules/structs/repo.go
Comment (lafriks) [reply modules/structs/repo.go]: Column should be added using Sync2 but data update should be made using single update by setting new column to updated_unix column value

Comment (JakobDev) [reply modules/structs/repo.go]: Can you link to example Migration using this?

**Comment by yardenshoham** [comment]
Try to use the time elements introduced in #23988

File: models/migrations/v1_20/v253.go
Comment (lunny) [inline models/migrations/v1_20/v253.go]:
```go
_, err := x.Exec("UPDATE repository SET archived_unix = 0")
return err
```

File: modules/structs/repo.go
Comment (lunny) [reply modules/structs/repo.go]: https://github.com/go-gitea/gitea/pull/22664/files#r1171375981

**Comment by JakobDev** [comment]
> Try to use the time elements introduced

I'm not sure if this fits well, as it works with relative time. This could be cause Problems in some Languages, as this is used within a sentence. I also think we should only show the Date.

File: modules/timeutil/timestamp.go
Comment (silverwind) [reply modules/timeutil/timestamp.go]: fine for now.

**Comment by silverwind** [comment]
@JakobDev please resolve any open resolved discussions so it's easier for reviewers to track the status.

**Comment by yardenshoham** [comment]
See https://github.com/go-gitea/gitea/pull/24275 to render the date

**Comment by JakobDev** [comment]
@yardenshoham I have now this line
```
<span id="archive-date-text"}>{{.locale.Tr "repo.archive.title_date" (DateTime "short" .Repository.ArchivedUnix.FormatLong .Repository.ArchivedUnix.FormatShort)}}</span
```
The Problem is, that your function outputs HTML, what `.locale.Tr` doesn't seem to like, as it gets' escaped.

**Comment by JakobDev** [comment]
> please resolve any open resolved discussions so it's easier for reviewers to track the status.

Should be done now

**Comment by silverwind** [comment]
Hmm I still see 5 open discussion threads.

**Comment by JakobDev** [comment]
> Hmm I still see 5 open discussion threads.

Forgot to close 2. The others are still open. 2 about Migration and 1 about the JS, which will probably removed with the new date function.

**Comment by delvh** [comment]
> @yardenshoham I have now this line
> 
> ```
> <span id="archive-date-text"}>{{.locale.Tr "repo.archive.title_date" (DateTime "short" .Repository.ArchivedUnix.FormatLong .Repository.ArchivedUnix.FormatShort)}}</span
> ```
> 
> The Problem is, that your function outputs HTML, what `.locale.Tr` doesn't seem to like, as it gets' escaped.

Pass it into `Safe` like so:
`{{.locale.Tr "<key>" <args, if present>  | Safe}}`

File: models/migrations/v1_20/v255.go
Comment (delvh) [inline models/migrations/v1_20/v255.go]:
```suggestion
		ArchivedUnix timeutil.TimeStamp `xorm:"DEFAULT 0"`
```
Non-caps-lock xorm commands scare me.

File: models/migrations/v1_20/v255.go
Comment (delvh) [inline models/migrations/v1_20/v255.go]: I don't quite understand the purpose of this statement.
That should already be done through the `DEFAULT`.

Or do you perhaps want to set it to `now` for all archived repos that don't have an archival time yet?

**Comment by JakobDev** [comment]
Thanks. We probably should document this somewhere.

File: models/migrations/v1_20/v255.go
Comment (JakobDev) [reply models/migrations/v1_20/v255.go]: This was [posted by lunny](https://github.com/go-gitea/gitea/pull/22664#discussion_r1171375981l): I also don't see why we need this, but perhaps some databases have Problem without running Update.

> Or do you perhaps want to set it to `now` for all archived repos that don't have an archival time yet?

That would result in a lot of incorrect data and confusion (Why was this Repo archived in 2023, but the last commit was in 2019? What happened in the meantime? Maybe some commits gets removed with a force push?), so I don't think this is a good idea. If we don't know the Date, we should not show a Date instead of a random wrong Date.

File: models/migrations/v1_20/v253.go
Comment (delvh) [reply models/migrations/v1_20/v253.go]: Why? I don't understand the purpose.
How is it better than a normal `Sync`?
Shouldn't this be done automatically through the `DEFAULT 0`?

File: models/migrations/v1_20/v255.go
Comment (delvh) [reply models/migrations/v1_20/v255.go]: Then I'm confused by the line as I wrote in lunny's comment above.

**Review by delvh — APPROVED** [review]

File: models/migrations/v1_20/v255.go
Comment (lafriks) [reply models/migrations/v1_20/v255.go]: It should be: `... SET archived_unix = updated_unix WHERE is_archived=?`, true

Comment (wxiaoguang) [reply models/migrations/v1_20/v255.go]: Yes, agree with  lafriks

File: services/convert/repository.go
Comment (wxiaoguang) [inline services/convert/repository.go]: I guess this `AsOptionalTime` is not the same behavior as other "time" related fields in Gitea.

In history, Gitea uses "0" / "1970-01-01" for no-time. But `AsOptionalTime` would get null.

If there is no strong reason, should we keep the old behavior for consistency?

Comment (JakobDev) [reply services/convert/repository.go]: I think it' better to return null, to indicate we have no Time instead of returning "1970-01-01".

Comment (wxiaoguang) [reply services/convert/repository.go]: It really causes inconsistency. Before all clients already use  "1970-01-01" for no-date. Now, the client sometimes needs to process null, sometimes needs to process  "1970-01-01" .....

Unless we have a full plan to refactor all date to support null.

File: templates/repo/home.tmpl
Comment (yardenshoham) [inline templates/repo/home.tmpl]:
```suggestion
					{{.locale.Tr "repo.archive.title_date" (DateTime "long" .Repository.ArchivedUnix.FormatLong .Repository.ArchivedUnix.FormatShort) | Safe}}
```

Show `April` instead of `Apr`

File: models/migrations/v1_20/v255.go
Comment (JakobDev) [reply models/migrations/v1_20/v255.go]: `updated_unix` is 0 when the date is unknown or the Repo is not archived. It can never be NULL, so we don't need zhe where here.

Comment (delvh) [reply models/migrations/v1_20/v255.go]: Actually, @lafriks' SQL is false, it needs to be
```sql
UPDATE repository SET archived_unix = updated_unix WHERE is_archived = 1 AND archived_unix = 0;
```
as you otherwise overwrite existing archive dates.

Comment (JakobDev) [reply models/migrations/v1_20/v255.go]: If the Migration is run, the `archived_unix` column will be added, what means this Column has no other Values than 0.

Comment (delvh) [reply models/migrations/v1_20/v255.go]: No, you cannot assume that.
There are many possible reasons why you would need to downgrade and rerun the migration again, beginning with local deployment as a maintainer to test out features, or a broken upgrade.
Asymmetric migrations should be avoided.

Comment (JakobDev) [reply models/migrations/v1_20/v255.go]: Should we really set `archived_unix` as the latest Update Date or should we left it with 0? Is it guaranteed that the latest update date is also the archive date? Does archiving a Repo also change `updated_unix`?

Comment (delvh) [reply models/migrations/v1_20/v255.go]: `updated_unix` is always updated when the `repository` entry is saved.
In our example, archival sets `is_archived = true`, thus the entry is updated and `updated_unix` with it.
So yes, it's guaranteed.

Comment (JakobDev) [reply models/migrations/v1_20/v255.go]: `updated_unix` is always updated when the `repository` entry is saved.

If I would have known that I wouldn't have added a new column, but now we have to Column, so we could stay with it. The Command is now updated.

Comment (delvh) [reply models/migrations/v1_20/v255.go]: No, it is still needed.
You can still change the settings even for an archived repo, so without the new column persisting it would potentially be overridden.
For example, you archive your repo on `day 3`.
On `day 5`, you change the repository name.
Then your archival date would suddenly be `day 5` instead of `day 3`.

Comment (JakobDev) [reply models/migrations/v1_20/v255.go]: This means that the Date can be wrong. I'm more in favour of "If we don't know the exact Date, we should not assume a Date", but if the most of you wants this, i can stay with it.

**Review by wxiaoguang — APPROVED** [review]

File: models/migrations/v1_20/v255.go
Comment (wxiaoguang) [reply models/migrations/v1_20/v255.go]: Hmm, I agree with "This means that the Date can be wrong" now, so I am neutral for this change. So LGTM.

**Comment by delvh** [comment]
Please run `make fmt` (or push an empty commit), that seems to be why the CI fails.

**Comment by wxiaoguang** [comment]
> Please run `make fmt` (or push an empty commit), that seems to be why the CI fails.

IIRC maintainers with writer permission could restart the CI manually.

**Comment by silverwind** [comment]
> IIRC maintainers with writer permission could restart the CI manually.

Only via CLI. If branch is update, there is nothing I can do on GitHub UI to restart a CI.

**Comment by silverwind** [comment]
Will review later.

**Comment by silverwind** [comment]
Does not work properly when archiving an empty repo:

Screenshot 2023-04-24 at 20 09 53

Works on regular repo:

Screenshot 2023-04-24 at 20 10 24

**Comment by silverwind** [comment]
@JakobDev the `repo.archive.title` which you removed is still in use in these two locations:

```
templates/repo/empty.tmpl:10:						{{.locale.Tr "repo.archive.title"}}
templates/repo/diff/compare.tmpl:222:					{{.locale.Tr "repo.archive.title"}}
```

Please be more carful and always "grep" removed translations.

**Review by silverwind — CHANGES_REQUESTED** [review]
Broken translations need to be fixed.

**Comment by JakobDev** [comment]
> @JakobDev the `repo.archive.title` which you removed is still in use in these two locations:

Now done. Good that you found that. these location needs to changed anywhere to display the date there.

**Review by silverwind — APPROVED** [review]

**Comment by yardenshoham** [comment]
@JakobDev please update the description with up-to-date screenshots

**Comment by lunny** [comment]
need @lafriks 's approval

**Review by lafriks — APPROVED** [review]

**Comment by delvh** [comment]
Note for maintainers:
Postpone the merge for at most two days.
It would indeed be good if the screenshots were updated to the latest state.
Apart from that, this PR fulfills the requirement to be merged.

File: templates/repo/home.tmpl
Comment (wxiaoguang) [inline templates/repo/home.tmpl]: I found a conflict ... after the `DateTime` fix:

* #24317

These `DateTime` calls should be updated .....

Comment (yardenshoham) [reply templates/repo/home.tmpl]: That's not a conflict (yet)

**Comment by JakobDev** [comment]
> please update the description with up-to-date screenshots

As I'm currently not on my Dev machine, I had updated the Description with the Screenshot from silberwind. i hope that's OK.

File: templates/repo/home.tmpl
Comment (JakobDev) [reply templates/repo/home.tmpl]: The linked PR needs to be merged first before I can update it

File: options/locale/locale_en-US.ini
Comment (yardenshoham) [inline options/locale/locale_en-US.ini]: Why not stay with `archive.title`?

Comment (wxiaoguang) [reply options/locale/locale_en-US.ini]: Agree we had better to stay with old key, then translators do not need to re-translate this unchanged string.

File: templates/repo/home.tmpl
Comment (silverwind) [reply templates/repo/home.tmpl]: Merge this first then?

Comment (yardenshoham) [reply templates/repo/home.tmpl]: Now it's a conflict

File: options/locale/locale_en-US.ini
Comment (silverwind) [reply options/locale/locale_en-US.ini]: Yeah, this is a pointless rename of a translation key. Apparently it already takes months to get some languages updated, so we need to be much more careful around translations, re-use where possible etc.

File: templates/repo/home.tmpl
Comment (JakobDev) [reply templates/repo/home.tmpl]: I have now updated this PR

File: options/locale/locale_en-US.ini
Comment (JakobDev) [reply options/locale/locale_en-US.ini]: The change was requested in another Review, but I have now changed it back.

**Review by wxiaoguang — APPROVED** [review]

File: models/migrations/v1_20/v255.go
Comment (lunny) [inline models/migrations/v1_20/v255.go]: The error should be checked.

**Pull request merged**
